app/cmd/rpc/mesh/session.go

```diff
@@ -89,124 +87,494 @@ func (sn DispatchResponse) GetSupportedNodes() []string {
+			// reschedule this session check because the node is not still on the expected block
+			nodeSession.ReScheduleValidationTask(s, servicerPubKey)
+			return
+		}
+
+		result, statusCode, e := s.GetDispatch(nodeSession)
+
+		if e != nil {
+			logger.Error(e.Error())
+			// -5 = read body issue
+			// StatusOK = 200
+			// StatusUnauthorized = 401 - maybe after few retries node runner fix the issue and this will move? should we retry this?
+			if statusCode == -5 || statusCode == http.StatusOK || statusCode == http.StatusUnauthorized {
+				// this will re queue this.
+				nodeSession.ReScheduleValidationTask(s, servicerPubKey)
+			}
+			return
+		}
+
+		isSuccess := statusCode == 200
+		nodeSession.Validated = true // no mater result, this was checked among the fullNode
+
+		if isSuccess {
+			// dispatch response about session - across nodes
+			s.Dispatch = result.Dispatch
+			// node-session specific
+			remainingRelays, _ := result.RemainingRelays.Int64()
+			nodeSession.RemainingRelays = remainingRelays
+			if result.Error != nil {
+				nodeSession.IsValid = !ShouldInvalidateSession(result.Error.Code)
+				if !nodeSession.IsValid {
+					nodeSession.Error = result.Error
+				}
+			} else {
+				nodeSession.IsValid = result.Success && remainingRelays > 0
+			}
+		} else if result.Error != nil {
+			nodeSession.IsValid = !ShouldInvalidateSession(result.Error.Code)
+		} else {
+			nodeSession.ReScheduleValidationTask(s, ServicerSessionEndpoint)
+		}
+	}
+}
+
+func (s *Session) GetDispatch(nodeSession *NodeSession) (result *RPCSessionResult, statusCode int, e error) {
+	servicerAddress, _ := GetAddressFromPubKeyAsString(nodeSession.PubKey)
+
+	servicerNode := getServicerFromPubKey(nodeSession.PubKey)
+
 	payload := pocketTypes.MeshSession{
 		SessionHeader: pocketTypes.SessionHeader{
-			ApplicationPubKey:  relay.Proof.Token.ApplicationPublicKey,
-			Chain:              relay.Proof.Blockchain,
-			SessionBlockHeight: relay.Proof.SessionBlockHeight,
+			ApplicationPubKey:  s.AppPublicKey,
+			Chain:              s.Chain,
+			SessionBlockHeight: s.BlockHeight,
 		},
-		Meta:               relay.Meta,
-		ServicerPubKey:     relay.Proof.ServicerPubKey,
-		Blockchain:         relay.Proof.Blockchain,
-		SessionBlockHeight: relay.Proof.SessionBlockHeight,
+		Meta:               *nodeSession.RelayMeta,
+		ServicerPubKey:     nodeSession.PubKey,
+		Blockchain:         s.Chain,
+		SessionBlockHeight: s.BlockHeight,
 	}
-	logger.Debug(fmt.Sprintf("reading session from servicer %s", servicerNode.Address.String()))
-	jsonData, e := json.Marshal(payload)
-	if e != nil {
-		return NewSdkErrorFromPocketSdkError(sdk.ErrInternal(e.Error()))
+
+	logger.Debug(
+		fmt.Sprintf(
+			"session store - reading session for app=%s chain=%s blockHeight=%d servicer=%s",
+			s.AppPublicKey,
+			s.Chain,
+			s.BlockHeight,
+			servicerAddress,
+		),
+	)
+	jsonData, e1 := json.Marshal(payload)
+	if e1 != nil {
+		statusCode = -2
```

Comment:
custom status codes are hard to maintain if you just do -1 to -5, use `iota`

https://www.gopherguides.com/articles/how-to-use-iota-in-golang
```
const (
	Red int = iota
	Orange
	Yellow
	Green
	Blue
	Indigo
	Violet
)
```

Reply:
100% thanks for the link, will implement asap.